### PR TITLE
ref: Move shared schema code out of ClickHouse module

### PR DESCRIPTION
**Pull request by lynnagara**
This change moves the shared column definitions from
`snuba.clickkhouse.columns` to `snuba.utils.schemas`. We will soon
be using these same column definitions for entities as well, and
we do not want to depend on the ClickHouse module as it is likely
to cause circular dependency issues.

The majority of this code has not changed, it is just copy and paste.

The ClickHouse column set now inherits from a base column set (which
currently does nothing). The same base class will also be used by
the entity column set class in the future.

### CODE REVIEW

File: snuba/clickhouse/columns.py [hunk 1]
```diff
@@ -1,475 +1,70 @@
...
+    "TModifiers",
+    "TypeModifier",
+    "TypeModifiers",
+    "UInt",
+    "UUID",
+)
 
 
-class ColumnSet:
+class ColumnSet(BaseColumnSet):
```
Comment (lynnagara) [inline snuba/clickhouse/columns.py:67]: This is the only part that changed, we are inheriting from a BaseColumnSet that does nothing.

**Review by evanh — APPROVED** [review]

File: snuba/clickhouse/columns.py [hunk 2]
```diff
@@ -1,475 +1,70 @@
...
+    String,
+    TModifiers,
+    TypeModifier,
+    TypeModifiers,
+    UInt,
+)
 
-    def get_raw(self) -> Enum[TModifiers]:
-        return Enum(self.values)
+__all__ = (
```
[the remark this answers is not in the document]
Comment (lynnagara) [reply snuba/clickhouse/columns.py:38]: Pretty much, just didn't want to rewrite all the imports right now

**Review by vladkluev — APPROVED** [review]
Thanks for the change. I just had some questions about the copy and pasted stuff (which I know you didn't write)

File: snuba/utils/schemas.py [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+        )
+
+    def _for_schema_impl(self) -> str:
+        return "Array({})".format(self.inner_type.for_schema())
+
+    def set_modifiers(self, modifiers: Optional[TModifiers]) -> Array[TModifiers]:
+        return Array(inner_type=self.inner_type, modifiers=modifiers)
+
+    def get_raw(self) -> Array[TModifiers]:
+        return Array(inner_type=self.inner_type.get_raw())
```
[the remark this answers is not in the document]
Comment (lynnagara) [reply snuba/utils/schemas.py:248]: It returns the same thing (I guess the "raw" type) but without any modifiers on it. I think was useful for some comparisons in our codebase, though I wonder if we really need it at all.

File: snuba/utils/schemas.py [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+        return "{} {}".format(escape_identifier(self.name), self.type.for_schema())
+
+    @staticmethod
+    def to_columns(
+        columns: Sequence[Union[Column[TModifiers], Tuple[str, ColumnType[TModifiers]]]]
+    ) -> Sequence[Column[TModifiers]]:
+        return [Column(*col) if not isinstance(col, Column) else col for col in columns]
+
+
+class FlattenedColumn:
```
[the remark this answers is not in the document]
Comment (lynnagara) [reply snuba/utils/schemas.py:152]: I think a flattened column is essentially very similar to a regular column but it also flattens out the `Nested` types so:

```
[Column("measurements", Nested([("key", String()), ("value", Float(64))]))]
```

becomes something like

```
[
    FlattenedColumn("measurements.key", String()),
    FlattenedColumn("measurements.value", Float(64))
]
```

This sort of mirrors "nested" columns in ClickHouse which are really just syntactic sugar (there is no actual thing as a nested column in a table)

File: snuba/utils/schemas.py [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+    Special type to be used in ColumnSets that represent the select
+    statement of a subquery since that is the schema the subquery
+    provides to the query that contains it.
+
+    Of course this cannot be used for migrations since this type does
+    not exist in Clickhouse.
+    """
+
+    def __init__(self) -> None:
+        super().__init__(None)
```
[the remark this answers is not in the document]
Comment (lynnagara) [reply snuba/utils/schemas.py:221]: The only place I can see it used is here: https://github.com/getsentry/snuba/blob/master/snuba/query/__init__.py#L132

Though I don't think that makes anything less confusing.

[the remark this answers is not in the document]
Comment (vladkluev) [reply snuba/utils/schemas.py:221]: Good to know I'm not crazy

**Pull request merged**